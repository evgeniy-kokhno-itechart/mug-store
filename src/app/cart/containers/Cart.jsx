import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import CartTableConnected from './CartTable';
import TotalCost from '../components/TotalCost';
import { cartCostsSelector } from '../CartSelectors';

export class Cart extends Component {
  renderCartDetails(products) {
    const { currentUserName, cart } = this.props;
    return (
      <React.Fragment>
        <CartTableConnected
          sortColumn="id"
          productsInCart={products}
          onDeleteFromCart={this.handleDeleteFromCart}
          onIncrementClick={this.handleIncrementClick}
          onDecrementClick={this.handleDecrementClick}
          onQuantityChange={this.handleQuantityChange}
        />
        <div className="row justify-content-between mx-2">
          <TotalCost total={cart.totalCost} customClasses="h5" />
          <div>
            <Link className="btn btn-secondary justify-content-end" to={currentUserName ? '/order/mycart' : '/order'}>
              Order Now!
            </Link>
          </div>
        </div>
      </React.Fragment>
    );
  }

  renderEmptyMessage() {
    return <h1 className="m-3 text-center">Your cart is empty</h1>;
  }

  render() {
    const { cart } = this.props;

    return cart.products.length ? this.renderCartDetails(cart.products) : this.renderEmptyMessage();
  }
}

Cart.propTypes = {
  cart: PropTypes.shape({
    products: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        imageURL: PropTypes.string,
        title: PropTypes.string,
        description: PropTypes.string,
        currentCurrencyPrice: PropTypes.number,
        quantity: PropTypes.number,
      }),
    ),
    totalCost: PropTypes.number,
  }).isRequired,

  currentUserName: PropTypes.string,
};

Cart.defaultProps = {
  currentUserName: '',
};

const mapStateToProps = state => ({
  cart: cartCostsSelector(state),
  currentUserName: state.user.currentUser.name,
});

export default connect(mapStateToProps)(Cart);
