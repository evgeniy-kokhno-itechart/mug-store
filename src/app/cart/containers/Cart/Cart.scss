.cart {
    margin: 1rem 0;

    &__empty-message {
        text-align: center;
    }
}

.cart-footer {
    display: flex;
    flex-wrap: wrap;
    margin: 1rem 1rem;
    white-space: nowrap;
    
    &__total {
        flex: 1;
        font-size: 1.5rem;
        font-weight: 500;
    }

    &__order-button {
        max-width: 8rem;
        flex: 1;
        justify-content: flex-end;
    }
}