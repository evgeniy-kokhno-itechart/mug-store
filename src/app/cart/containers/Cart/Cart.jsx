import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { PropTypes } from 'prop-types';
import CartTable from '../../components/CartTable/CartTable';
import TotalCost from '../../components/TotalCost';
import { cartCostsSelector } from '../../CartSelectors';
import { cartActions } from '../../CartActions';
import './Cart.scss';

export class Cart extends Component {
  renderCartDetails(products) {
    const {
      currentUserName, cart, isCurrencyLoading, changeQuantity, deleteProductFromCart,
    } = this.props;
    return (
      <React.Fragment>
        <CartTable
          sortColumn="id"
          productsInCart={products}
          isCurrencyLoading={isCurrencyLoading}
          changeQuantity={changeQuantity}
          deleteProductFromCart={deleteProductFromCart}
        />
        <div className="cart-footer">
          <TotalCost total={cart.totalCost} customClasses="cart-footer__total" />
          <Link className="button button--solid cart-footer__order-button" to={currentUserName ? '/order/mycart' : '/order'}>
            Order Now!
          </Link>
        </div>
      </React.Fragment>
    );
  }

  renderEmptyMessage() {
    return <h1 className="cart__empty-message">Your cart is empty</h1>;
  }

  render() {
    const { cart } = this.props;

    return (
      <div className="cart">
        {cart.products.length
          ? this.renderCartDetails(cart.products)
          : this.renderEmptyMessage()}
      </div>
    );
  }
}

Cart.propTypes = {
  cart: PropTypes.shape({
    products: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.string,
        imageURL: PropTypes.string,
        title: PropTypes.string,
        description: PropTypes.string,
        currentCurrencyCost: PropTypes.number,
        quantity: PropTypes.number,
      }),
    ),
    totalCost: PropTypes.number,
  }).isRequired,

  isCurrencyLoading: PropTypes.bool,

  currentUserName: PropTypes.string,

  changeQuantity: PropTypes.func.isRequired,
  deleteProductFromCart: PropTypes.func.isRequired,
};

Cart.defaultProps = {
  isCurrencyLoading: true,
  currentUserName: '',
};

const mapStateToProps = state => ({
  cart: cartCostsSelector(state),
  currentUserName: state.user.currentUser.name,
  isCurrencyLoading: state.currency.currenciesStatus.isInProcess,
});

const mapDispatchToProps = {
  changeQuantity: cartActions.ChangeQuantity,
  deleteProductFromCart: cartActions.DeleteFromCart,
};

export default connect(mapStateToProps, mapDispatchToProps)(Cart);
