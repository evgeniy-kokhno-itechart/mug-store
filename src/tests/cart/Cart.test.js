/* eslint-disable no-undef */
import React from 'react';

import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import { Cart } from '../../app/cart/containers/Cart/Cart';

configure({ adapter: new Adapter() });

describe('<Cart />', () => {
  let CartWrapper;
  let changeQuantity;
  let deleteProductFromCart;


  beforeEach(() => {
    changeQuantity = jest.fn();
    deleteProductFromCart = jest.fn();

    CartWrapper = shallow(
      <Cart
        cart={{ products: [], totalCost: 0 }}
        changeQuantity={changeQuantity}
        deleteProductFromCart={deleteProductFromCart}
      />,
    );
  });

  it('renders properly with default props', () => {
    expect(CartWrapper).toMatchSnapshot();
  });

  it('renders properly with cart array not empty', () => {
    CartWrapper.setProps({
      cart: {
        products: [
          {
            id: '1',
            imageURL: '#',
            title: 'test product 1',
            description: 'test product 1 description',
            currentCurrencyCost: 5,
            quantity: 3,
          },
          {
            id: '2',
            imageURL: '#',
            title: 'test product 2',
            description: 'test product 2 description',
            currentCurrencyCost: 10,
            quantity: 5,
          },
        ],
        totalCost: 15,
      },
    });
    expect(CartWrapper).toMatchSnapshot();
  });

  it('renders properly with not empty cart and currentUserName provided', () => {
    CartWrapper.setProps({
      cart: {
        products: [
          {
            id: '1',
            imageURL: '#',
            title: 'test product 1',
            description: 'test product 1 description',
            currentCurrencyCost: 5,
            quantity: 3,
          },
        ],
        totalCost: 5,
      },
      currentUserName: 'test user',
    });
    expect(CartWrapper).toMatchSnapshot();
  });
});
