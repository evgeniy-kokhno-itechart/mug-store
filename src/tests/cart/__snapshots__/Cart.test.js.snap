// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart /> renders properly with cart array not empty 1`] = `
<div
  className="cart"
>
  <CartTable
    changeQuantity={[MockFunction]}
    deleteProductFromCart={[MockFunction]}
    isCurrencyLoading={true}
    productsInCart={
      Array [
        Object {
          "currentCurrencyCost": 5,
          "description": "test product 1 description",
          "id": "1",
          "imageURL": "#",
          "quantity": 3,
          "title": "test product 1",
        },
        Object {
          "currentCurrencyCost": 10,
          "description": "test product 2 description",
          "id": "2",
          "imageURL": "#",
          "quantity": 5,
          "title": "test product 2",
        },
      ]
    }
    sortColumn="id"
  />
  <div
    className="cart-footer"
  >
    <TotalCost
      customClasses="cart-footer__total"
      total={15}
    />
    <Link
      className="button button-solid cart-footer__order-button"
      to="/order"
    >
      Order Now!
    </Link>
  </div>
</div>
`;

exports[`<Cart /> renders properly with default props 1`] = `
<div
  className="cart"
>
  <h1
    className="cart__empty-message"
  >
    Your cart is empty
  </h1>
</div>
`;

exports[`<Cart /> renders properly with not empty cart and currentUserName provided 1`] = `
<div
  className="cart"
>
  <CartTable
    changeQuantity={[MockFunction]}
    deleteProductFromCart={[MockFunction]}
    isCurrencyLoading={true}
    productsInCart={
      Array [
        Object {
          "currentCurrencyCost": 5,
          "description": "test product 1 description",
          "id": "1",
          "imageURL": "#",
          "quantity": 3,
          "title": "test product 1",
        },
      ]
    }
    sortColumn="id"
  />
  <div
    className="cart-footer"
  >
    <TotalCost
      customClasses="cart-footer__total"
      total={5}
    />
    <Link
      className="button button-solid cart-footer__order-button"
      to="/order/mycart"
    >
      Order Now!
    </Link>
  </div>
</div>
`;
